File: 46-permutations/permutations.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        res = []

        def dfs(i, temp_nums, curr):
            # insert ith element
            new_curr = curr + [temp_nums[i]]
            new_temp_nums = temp_nums[:i] + temp_nums[i+1:]

            # base case: no elements left
            if len(new_temp_nums) == 0:
                res.append(new_curr)
                return

            for j in range(len(new_temp_nums)):
                dfs(j, new_temp_nums, new_curr)

        # Start recursion with each possible first element
        for i in range(len(nums)):
            dfs(i, nums, [])
        
        return res

File: 46-permutations/permutations.cpp
class Solution {
private:
    void dfs(int i, vector<int> temp_nums, vector<int> curr, vector<vector<int>> &res){
        curr.push_back(temp_nums[i]);
        temp_nums.erase(temp_nums.begin() + i);

        if(temp_nums.size() == 0){
            res.push_back(curr);
            return;
        }

        for(int j=0; j<temp_nums.size(); j++){
            dfs(j, temp_nums, curr, res);
        }
    }
public:
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> res;

        for(int i=0; i<nums.size(); i++){
            dfs(i, nums, {}, res);
        }

        return res;
    }
};